astropy/units/core.py

```diff
@@ -461,6 +461,11 @@ class UnitConversionError(UnitsError, ValueError):
     interpreting units in terms of other units.
     """
 
+class UnitTypeError(UnitsError, TypeError):
+    """
+    Used specifically for errors in setting to units not allowed by a class.
```

Comment:
The only reason I made this is that currently `angle *= u.m` would raise a `TypeError` (in `Quantity.__array_prepare__`, and I felt this was sufficiently similar to `angle.view(Distance)` that it should have the same type of exception.  I realise now, though, that I didn't actually make the change in `__array_prepare__`; will do that now...